packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts

```diff
@@ -222,11 +327,91 @@ function rebaseMark<TNodeChange>(
             }
             return clone(currMark);
         }
+        case "MoveOut":
+        case "ReturnFrom": {
+            if (!isSkipMark(currMark)) {
+                getOrAddEmptyToMap(moveEffects.movedMarks, baseMark.id).push(clone(currMark));
+            }
+            return 0;
+        }
         default:
             fail(`Unsupported mark type: ${baseType}`);
     }
 }
 
+function applyMoveEffects<TNodeChange>(
+    baseMarks: MarkList<TNodeChange>,
+    rebasedMarks: MarkList<TNodeChange>,
+    moveEffects: MoveEffectTable<TNodeChange>,
+    genId: IdAllocator,
+): Changeset<TNodeChange> {
+    const queue = new ComposeQueue<TNodeChange>(
+        baseMarks,
+        undefined,
+        rebasedMarks,
+        () => fail("Should not split moves while applying move effects"),
+        moveEffects,
+        false,
+    );
+    const factory = new MarkListFactory<TNodeChange>(moveEffects);
+
+    let offset = 0;
+    while (!queue.isEmpty()) {
+        const { baseMark, newMark } = queue.pop();
+        if (isObjMark(baseMark) && (baseMark.type === "MoveIn" || baseMark.type === "ReturnTo")) {
+            const movedMarks = moveEffects.movedMarks.get(baseMark.id);
+            if (movedMarks !== undefined) {
+                factory.pushOffset(offset);
+                offset = 0;
+
+                // TODO: Do moved marks ever need to be split?
+                factory.push(...movedMarks);
+                const size = movedMarks.reduce<number>(
+                    (count, mark) => count + getInputLength(mark),
+                    0,
+                );
+                factory.pushOffset(-size);
+            }
+        }
+        if (newMark === undefined) {
+            assert(baseMark !== undefined, "Non-empty RebaseQueue should return at least one mark");
+            offset += getOutputLength(baseMark);
+            continue;
+        }
+
+        // TODO: Offset wouldn't be needed if queue returned skip instead of undefined in cases where it should return two marks
+        offset = 0;
+        if (isObjMark(newMark)) {
+            switch (newMark.type) {
+                case "MoveIn": {
```

Comment:
Yes. I was wondering why this isn't caught by the rebase tests, but I realized it's because this whole block is now redundant. `ComposeQueue.pop` splits the marks before returning them.